### PR TITLE
Feature/offensive positioning

### PR DESCRIPTION
**Pull request by RobZelluf**
- [x] The tests are passing

Work in progress

### Summary
Used to have side attackers find the optimal position to stand idle

### CODE REVIEW

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 1]
```diff
@@ -1,3 +1,7 @@
+#include <utility>
+
+#include <utility>
```
Comment (Barisimre) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: /r/hmmm

File: roboteam_ai/src/skills/SideAttacker.cpp [hunk 2]
```diff
@@ -3,40 +3,86 @@
 //
 
 #include "SideAttacker.h"
-
+#include <chrono>
```
Comment (Barisimre) [inline roboteam_ai/src/skills/SideAttacker.cpp]: Remove if you not timing things

File: roboteam_ai/src/skills/SideAttacker.cpp [hunk 3]
```diff
@@ -3,40 +3,86 @@
...
+
+        for (auto &i : robotsPositioning) {
+            robotLocations.emplace_back(i->pos);
+        }
+
+        auto shortestDistances = control::ControlUtils::calculateClosestPathsFromTwoSetsOfPoints(robotLocations,
+                                                                                                 targetLocations);
+
+        for (unsigned long i = 0; i < robotsPositioning.size(); i++) {
+            if (robotsPositioning.at(i)->id == robot->id) {
```
Comment (Barisimre) [inline roboteam_ai/src/skills/SideAttacker.cpp]: Maybe try iterator loop instead of range based

File: roboteam_ai/src/skills/Skill.h [hunk 4]
```diff
@@ -28,7 +28,7 @@ class Skill : public bt::Leaf {
     io::IOManager ioManager = io::IOManager(false, true);
     void publishRobotCommand();
     void refreshRobotCommand();
-    roboteam_msgs::RobotCommand command;
+    roboteam_msgs::RobotCommand command; //TODO: Does this properly work?
```
Comment (Barisimre) [inline roboteam_ai/src/skills/Skill.h]: idk does it

File: roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    int counter = 0;
+    std::vector<Vector2> newOffensivePositions;
+
+    // Calculate the next offensive positions 10 times
+    while (counter < 10) {
+        newOffensivePositions = rtt::ai::coach::g_offensiveCoach.getNewOffensivePositions();
+        counter++;
+    }
+
+    for(int i = 0; i < offensivePositions.size(); i++) {
```
Comment (Barisimre) [inline roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp]: Try iterator loop again maybe

**Review by Barisimre — CHANGES_REQUESTED** [review]

File: roboteam_ai/src/skills/SideAttacker.cpp (hunk 3 above)
Comment (RobZelluf) [reply roboteam_ai/src/skills/SideAttacker.cpp]: No

File: roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp (hunk 5 above)
Comment (RobZelluf) [reply roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp]: No

File: roboteam_ai/src/skills/Skill.h (hunk 4 above)
Comment (RobZelluf) [reply roboteam_ai/src/skills/Skill.h]: It does

File: roboteam_ai/src/skills/SideAttacker.cpp (hunk 2 above)
Comment (RobZelluf) [reply roboteam_ai/src/skills/SideAttacker.cpp]: Done

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 6]
```diff
@@ -75,36 +62,33 @@ double CoachHeuristics::calculateDistanceFromCornerScore(const Vector2 &position
...
-    auto w = world::world->getWorld();
-    auto ball = world::world->getBall();
+double CoachHeuristics::calculatePositionScore(const Vector2& position) {
+    WorldData world = world::world->getWorld();
+    roboteam_msgs::GeometryFieldSize field = world::field->get_field();
+    double closeToGoalScore = calculateCloseToGoalScore(position);
+    double shotAtGoalScore = calculateShotAtGoalScore(position, world);
 
-    if (!ball) return -1.0;
+    double score = 3 * shotAtGoalScore + closeToGoalScore;
```
Comment (Barisimre) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: hah got you there

**Review by Barisimre — APPROVED** [review]

File: roboteam_ai/src/coach/CoachHeuristics.h [hunk 7]
```diff
@@ -15,32 +15,33 @@ namespace ai {
...
+    static const double SHOT_AT_GOAL_WEIGHT;
+    static const double PASS_LINE_WEIGHT;
+    static const double DISTANCE_TO_OPPONENTS_WEIGHT;
+    static const double DISTANCE_FROM_CORNER_WEIGHT;
+public:
+    static double calculateCloseToGoalScore(const Vector2& position);
+    static double calculateShotAtGoalScore(const Vector2& position, WorldData world);
+    static double calculatePassLineScore(const Vector2& position, WorldData world);
+    static double calculateDistanceToOpponentsScore(const Vector2 &position, const WorldData world);
+    static double calculateDistanceFromCornerScore(const Vector2& position, const roboteam_msgs::GeometryFieldSize& field);
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/coach/CoachHeuristics.h]: preferably for references &  : Vector2 &position or Vector2 & position
and for pointers *                  : Vector2* position or Vector2 * position

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 8]
```diff
@@ -82,15 +83,31 @@ bool ControlUtils::clearLine(Vector2 fromPos, Vector2 toPos, world::WorldData wo
     return true;
 }
 
+double ControlUtils::closestEnemyToLineDistance(const Vector2 &fromPos, Vector2 toPos, const world::WorldData world,
+                                                bool keeper) {
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/control/ControlUtils.cpp]: fix this line

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 9]
```diff
@@ -82,15 +83,31 @@ bool ControlUtils::clearLine(Vector2 fromPos, Vector2 toPos, world::WorldData wo
...
+    double shortestDistance = INT_MAX;
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/control/ControlUtils.cpp]: double =!=!= int

File: roboteam_ai/src/skills/SideAttacker.cpp [hunk 10]
```diff
@@ -7,36 +7,82 @@
...
+    robotsPositioning.emplace_back(robot);
 }
 
 
 /// Get an update on the skill
 bt::Node::Status SideAttacker::onUpdate() {
-    if (! robot) return Status::Running;
-    targetPos = coach::g_offensiveCoach.calculatePositionForRobot(robot);
 
+    bool isIn = false;
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/skills/SideAttacker.cpp]: is in what? make a descriptive name

File: roboteam_ai/src/skills/SideAttacker.cpp [hunk 11]
```diff
@@ -7,36 +7,82 @@
...
+    for (auto & i : robotsPositioning) {
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/skills/SideAttacker.cpp]: for (auto &robotPositioning : robotsPositiong)

File: roboteam_ai/src/skills/SideAttacker.cpp [hunk 12]
```diff
@@ -7,36 +7,82 @@
...
+Vector2 SideAttacker::getOffensivePosition() {
+    auto field = world::field->get_field();
+
+    std::vector<Vector2> targetLocations = coach::g_offensiveCoach.getNewOffensivePositions();
+
+    // If not yet assigned to a zone, do so according to the Hungarian
+    if (zone == -1) {
+        std::vector<Vector2> robotLocations;
+
+        for (auto &i : robotsPositioning) {
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/src/skills/SideAttacker.cpp]: i

File: roboteam_ai/src/skills/SideAttacker.cpp (hunk 3 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/skills/SideAttacker.cpp]: yes

File: roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp (hunk 5 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/test/CoachTests/OffensiveCoachTest.cpp]: x

**Review by ThijsLuttikhuis — CHANGES_REQUESTED** [review]
some stuff could be in gameanalyzer, but can be done later
fix some names to make clear what their use is

**Review by ThijsLuttikhuis — APPROVED** [review]

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 13]
```diff
@@ -12,198 +12,48 @@ namespace coach {
...
+OffensiveCoach::OffensivePosition OffensiveCoach::calculateNewRobotPosition(const OffensivePosition& currentPosition, const Vector2& defaultPosition) {
 
-    position.position = {x, y};
+    OffensivePosition bestPosition = currentPosition;
+    bestPosition.score = CoachHeuristics::calculatePositionScore(bestPosition.position);
 
-    if (!Field::pointIsInField(position.position) || Field::pointIsInDefenceArea(position.position, false)) {
-        return calculateRandomPosition(xStart, xEnd, yStart, yEnd);
-    }
+    for (int xDiff = -GRID_SIZE; xDiff <= GRID_SIZE; xDiff++) {
```
[the remark this answers is not in the document]
Comment (mrlukasbos) [reply roboteam_ai/src/coach/OffensiveCoach.cpp]: just remove the `=` then

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 14]
```diff
@@ -24,34 +24,21 @@ double CoachHeuristics::calculateCloseToGoalScore(const Vector2 &position) {
...
-        if (control::ControlUtils::clearLine(position, world::field->get_their_goal_center(), world, safeDistanceFactor)) {
-            break;
-        }
-        safeDistanceFactor -= 0.5;
-    }
+double CoachHeuristics::calculateShotAtGoalScore(const Vector2& position, WorldData world) {
+    double shortestDistance = control::ControlUtils::closestEnemyToLineDistance(position, world::field->get_their_goal_center(), world, false);
 
-    return 1 - exp(SHOT_AT_GOAL_WEIGHT * safeDistanceFactor);
+    return 1 - exp(SHOT_AT_GOAL_WEIGHT * shortestDistance);
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: why the exp?

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 15]
```diff
@@ -24,34 +24,21 @@ double CoachHeuristics::calculateCloseToGoalScore(const Vector2 &position) {
...
-        if (control::ControlUtils::clearLine(world.ball.pos, position, world, safeDistanceFactor)) {
-            break;
-        }
-        safeDistanceFactor -= 0.5;
-    }
+double CoachHeuristics::calculatePassLineScore(const Vector2& position, WorldData world) {
+    double shortestDistance = control::ControlUtils::closestEnemyToLineDistance(world.ball.pos, position, world, false);
 
-    return 1 - exp(PASS_LINE_WEIGHT * safeDistanceFactor);
+    return 1 - exp(PASS_LINE_WEIGHT * shortestDistance);
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: exp?

File: roboteam_ai/src/coach/CoachHeuristics.cpp (hunk 6 above)
Comment (mrlukasbos) [reply roboteam_ai/src/coach/CoachHeuristics.cpp]: magic number

File: roboteam_ai/src/coach/CoachHeuristics.h [hunk 16]
```diff
@@ -15,32 +15,35 @@ namespace ai {
...
-                BallPtr &ball);
-        static double calculatePositionScore(const Vector2 &position);
+private:
+    using WorldData = world::WorldData;
+    using Ball = world::Ball;
+    using Robot = world::Robot;
+    using BallPtr = world::World::BallPtr;
+    using RobotPtr = world::World::RobotPtr;
+
+    static const double MAX_DISTANCE_FROM_BALL;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.h]: why are these static? Why not instantly initialize them?

Comment (mrlukasbos) [reply roboteam_ai/src/coach/CoachHeuristics.h]: const means static anyway

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 17]
```diff
@@ -12,198 +12,48 @@ namespace coach {
...
+        if (currentPosition.position.x < 0 && xDiff <= 0) continue;
 
-    return position;
-}
+        for (int yDiff = -GRID_SIZE; yDiff <= GRID_SIZE; yDiff++) {
+            OffensivePosition newPosition;
```
[the remark this answers is not in the document]
Comment (mrlukasbos) [reply roboteam_ai/src/coach/OffensiveCoach.cpp]: yes

**Review by mrlukasbos — APPROVED** [review]
yeah its okay. I like the chosen positions very much

**Pull request merged**